memory accounting - reduce code in non DEBUG build

mykaul opened the pull request:
Subset of https://github.com/gluster/glusterfs/pull/3852 - just the changes for the memory accounting parts:
- reorder some variables, according to access pattern
- ifdef code that's only relevant under DEBUG mode.

Commit:
mem-pool minor changes

- gf_mem_update_acct_info() is not needed unless compiling in DBEUG
- Change the order of some variables for better access pattern.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
xlator.c: memory account records initialization should be done only under DEBUG compilation

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/mem-pool.c b/libglusterfs/src/mem-pool.c
index 084caac36..2b8fbfd22 100644
--- a/libglusterfs/src/mem-pool.c
+++ b/libglusterfs/src/mem-pool.c
@@ -306,8 +306,8 @@ __gf_mem_invalidate(void *ptr)
 
     /* zero out remaining (old) mem_header bytes) */
     /* and the first byte of data */
-    memset(old_ptr + sizeof(inval), 0x00, (sizeof(struct mem_header) - sizeof(inval)) + 1);
-
+    memset(old_ptr + sizeof(inval), 0x00,
+           (sizeof(struct mem_header) - sizeof(inval)) + 1);
 }
 #endif /* DEBUG */
 
```

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/mem-pool.c b/libglusterfs/src/mem-pool.c
index 084caac36..2b8fbfd22 100644
--- a/libglusterfs/src/mem-pool.c
+++ b/libglusterfs/src/mem-pool.c
@@ -306,8 +306,8 @@ __gf_mem_invalidate(void *ptr)
 
     /* zero out remaining (old) mem_header bytes) */
     /* and the first byte of data */
-    memset(old_ptr + sizeof(inval), 0x00, (sizeof(struct mem_header) - sizeof(inval)) + 1);
-
+    memset(old_ptr + sizeof(inval), 0x00,
+           (sizeof(struct mem_header) - sizeof(inval)) + 1);
 }
 #endif /* DEBUG */
 
diff --git a/libglusterfs/src/xlator.c b/libglusterfs/src/xlator.c
index affa41b3c..0d0e0bba1 100644
--- a/libglusterfs/src/xlator.c
+++ b/libglusterfs/src/xlator.c
@@ -758,7 +758,6 @@ xlator_mem_acct_init(xlator_t *xl, int num_types)
 void
 xlator_mem_acct_unref(struct mem_acct *mem_acct)
 {
-
     if (GF_ATOMIC_DEC(mem_acct->refcnt) == 0) {
 #ifdef DEBUG
         uint32_t i;
```

Commit:
clang-format fix

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/mem-pool.c b/libglusterfs/src/mem-pool.c
index 084caac36..2b8fbfd22 100644
--- a/libglusterfs/src/mem-pool.c
+++ b/libglusterfs/src/mem-pool.c
@@ -306,8 +306,8 @@ __gf_mem_invalidate(void *ptr)
 
     /* zero out remaining (old) mem_header bytes) */
     /* and the first byte of data */
-    memset(old_ptr + sizeof(inval), 0x00, (sizeof(struct mem_header) - sizeof(inval)) + 1);
-
+    memset(old_ptr + sizeof(inval), 0x00,
+           (sizeof(struct mem_header) - sizeof(inval)) + 1);
 }
 #endif /* DEBUG */
 
```

mykaul commented:
/run regression

mykaul commented on tests/basic/fop-sampling.t:
```diff
@@ -39,7 +39,7 @@ cleanup;
 TEST pidof glusterd
 TEST $CLI volume create $V0 replica 3 $H0:$B0/${V0}{0,1,2}
-TEST $CLI volume set $V0 nfs.disable off
+$CLI volume set $V0 nfs.disable off
```
This change is done so testing without gNFS won't fail

Commit:
Multiple changes to reduce overhead of memory accounting

- gf_mem_update_acct_info() is not needed when not in DEBUG mode
- re-order variables in the structure according to access pattern
- Turn xlator_mem_acct_unref() into xlator_mem_acct_destroy() and call it only when refcnt euqals 0 - which is quite rare.

Updates: #3855
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
/run regression
